File: R/Lecture 04.R

# STAT 521: Lecture 4

# Central limit theorem
# Require MASS package
library(MASS)

# Set up a hypothetical population
# (using an exponential distribution)
population <- rexp(10 ^ 4, 1 / 20) + 20

# Population distribution
truehist(population)
title('Population distribution (hypothetical)')

# Mean & SD of the population distribution
# Mean is around 40, SD is around 20
describe <- function(x) c(Mean = mean(x), SD = sd(x))
describe(population)

# Take a random sample of n = 40
n <- 40
samp1 <- sample(population, n)
truehist(samp1)
describe(samp1)

# Now take 1000 samples of n = 40
# and calculate a mean each time
samp.mean <- rep(NA, 1000)
samp.mean[] <- sapply(samp.mean, function(x) mean(sample(population, n)))
samp.mean

# Sampling distribution of means
truehist(samp.mean, nbins = 30)
lines(density(samp.mean), col = 'red', lwd = 2)
title('Sampling distribution of means')

# Look at the mean & SD of sample means
# Mean should be around 40 and SD around 20/sqrt(40) = 3.16
describe(samp.mean)

File: Python/lecture04.py

# STAT 521: Lecture 4

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from random import sample

# Central limit theorem
# Set up a hypothetical population
population = np.random.exponential(20, 10 ** 4) + 20

# Population distribution
ax = sns.displot(population)
ax.fig.suptitle("Population distribution (hypothetical)")
plt.show()
plt.clf()

# Mean & SD of the population distribution
# Mean is around 40, SD is around 20
df = pd.DataFrame({"population": population})
df.describe()

# Take a random sample of n = 40
n = 40
samp1 = sample(list(population), n)
sns.displot(samp1)
plt.show()
plt.clf()

df = pd.DataFrame({"sample 1": samp1})
df.describe()

# Now take 1000 samples of n = 40
n_samples = 1000
df2 = pd.DataFrame(index = range(1, n + 1))

for i in range(1, n_samples + 1):
    sample_i = sample(list(population), n)
    col = f"sample {i}"
    df2[col] = sample_i

# ...and calculate a mean from each sample
sample_means = df2.mean()
sample_means.head(n = 10)

# Sampling distribution of means
ax = sns.displot(sample_means, kde = True)
ax.fig.suptitle("Sampling distribution of means")
plt.show()
plt.clf()

# Look at the mean & SD of sample means
# Mean should be around 40 and SD around 20/sqrt(40) = 3.16
sample_means.describe()
